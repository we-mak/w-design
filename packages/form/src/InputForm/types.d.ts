import { ActionProps } from "@w-design/core";
import { InputFieldProps } from "../InputField/types";

export interface InputFormProps extends InputFieldProps, ActionProps {
  label?: string;
  isError?: boolean;
  isWarning?: boolean;
  isSuccess?: boolean;
  hintMessage?: string;
}
