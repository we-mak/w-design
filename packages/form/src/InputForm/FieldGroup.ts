import styled from "styled-components";
import { InputFormProps } from "./index";

export const FieldGroup = styled.div<InputFormProps>`
  position: relative;
  font-family: ${props => `${props.theme.fonts["normal"]}px`};
  &:not(:last-child) {
    margin-bottom: ${props => props.theme.space[2]}px;
  }

  i {
    color: ${props => props.theme.colors["N10"]};

    &.icon-before {
      height: 0.8rem;
      margin: 0 0.25rem;
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      width: 0.8rem;
      z-index: 2;
      left: 0.1rem;
    }
    &.icon-after {
      height: 0.8rem;
      margin: 0 0.25rem;
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      width: 0.8rem;
      z-index: 2;
      right: 0.1rem;
    }
  }
  input {
    &.icon-before {
      padding-left: 1.4rem !important;
    }
    &.icon-after {
      padding-right: 1.4rem !important;
    }
  }
`;
