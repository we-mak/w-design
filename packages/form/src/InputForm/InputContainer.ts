import styled from "styled-components";

export const InputContainer = styled.span`
  position: relative;
  display: block;
`;
