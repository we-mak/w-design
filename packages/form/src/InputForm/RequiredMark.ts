import styled from "styled-components";

export const RequiredMark = styled.sup`
  color: ${props => props.theme.colors["R30"]};
  font-size: ${props => props.theme.fontSizes[0]};
  top: -0.2rem;
`;
