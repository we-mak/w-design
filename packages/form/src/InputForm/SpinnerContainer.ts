import styled from "styled-components";

export const SpinnerContainer = styled.span`
  margin: 0 0.25rem;
  position: absolute;
  top: 10%;
  z-index: 2;
  right: 0;
  display: block;
`;
