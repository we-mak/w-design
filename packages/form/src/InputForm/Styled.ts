import styled from "styled-components";
import { InputFormProps } from "./index";

export const FieldGroup = styled.div<InputFormProps>`
  position: relative;
  font-family: ${props => `${props.theme.fonts["normal"]}px`};
  &:not(:last-child) {
    margin-bottom: ${props => props.theme.space[2]}px;
  }

  i {
    color: ${props => props.theme.colors["N10"]};

    &.icon-before {
      height: 0.8rem;
      margin: 0 0.25rem;
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      width: 0.8rem;
      z-index: 2;
      left: 0.1rem;
    }
    &.icon-after {
      height: 0.8rem;
      margin: 0 0.25rem;
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      width: 0.8rem;
      z-index: 2;
      right: 0.1rem;
    }
  }
  input {
    &.icon-before {
      padding-left: 1.4rem !important;
    }
    &.icon-after {
      padding-right: 1.4rem !important;
    }
  }
`;
FieldGroup.displayName = "FieldGroup";

export const InputContainer = styled.span`
  position: relative;
  display: block;
`;

export const TextHint = styled.small<InputFormProps>`
  display: block;
  margin-top: 0.4rem;
  color: ${props => {
    const { isWarning, isError, isSuccess, theme } = props;
    if (isWarning) return `${theme.colors["O30"]}`;
    if (isError) return `${theme.colors["R30"]}`;
    if (isSuccess) return `${theme.colors["G30"]}`;
    return `${theme.colors["N30"]}`;
  }};
`;
TextHint.displayName = "TextHint";

export const RequiredMark = styled.sup`
  color: ${props => props.theme.colors["R30"]};
  font-size: ${props => props.theme.fontSizes[0]};
  top: -0.2rem;
`;
RequiredMark.displayName = "RequiredMark";

export const SpinnerWrap = styled.span`
  margin: 0 0.25rem;
  position: absolute;
  top: 10%;
  z-index: 2;
  right: 0;
  display: block;
`;
SpinnerWrap.displayName = "SpinnerWrap";
