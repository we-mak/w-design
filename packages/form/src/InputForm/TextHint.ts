import styled from "styled-components";
import { InputFormProps } from "./index";

export const TextHint = styled.small<InputFormProps>`
  display: block;
  color: ${props => {
    const { isWarning, isError, isSuccess, theme } = props;
    if (isWarning) return `${theme.colors["O30"]}`;
    if (isError) return `${theme.colors["R30"]}`;
    if (isSuccess) return `${theme.colors["G30"]}`;
    return `${theme.colors["N30"]}`;
  }};
`;
