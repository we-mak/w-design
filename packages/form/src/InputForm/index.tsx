/**
 * Input form
 * Input form with label, text hint...
 */
import * as React from "react";
import { Spinner, Icon } from "@w-design/core";
import { ActionProps } from "@w-design/core/lib/types/common/props";
import Label from "../Label";
import InputField from "../InputField";
import { InputFieldProps } from "../InputField";
import { InputContainer } from "./InputContainer";
import { TextHint } from "./TextHint";
import { SpinnerContainer } from "./SpinnerContainer";
import { FieldGroup } from "./FieldGroup";
import { RequiredMark } from "./RequiredMark";

export interface InputFormProps extends InputFieldProps, ActionProps {
  label?: string;
  isError?: boolean;
  isWarning?: boolean;
  isSuccess?: boolean;
  hintMessage?: string;
}

const InputForm: React.FunctionComponent<InputFormProps> = props => {
  const {
    id,
    label,
    hintMessage,
    iconBefore,
    iconAfter,
    isLoading,
    isRequired,
    isError,
    isWarning,
    isSuccess,
    children,
    ...rest
  } = props;
  return (
    <FieldGroup>
      <Label htmlFor={id}>
        {label} {isRequired && <RequiredMark aria-hidden="true">*</RequiredMark>}
        <InputContainer>
          {iconBefore && <Icon className={`icon-before ${iconBefore}`} />}
          <InputField
            className={(iconBefore && `icon-before`) || (iconAfter && `icon-after`)}
            fluid
            id={id}
            isError={isError}
            isWarning={isWarning}
            isRequired={isRequired}
            isValid={isSuccess}
            {...rest}
          />

          {iconAfter && <Icon className={`icon-after ${iconAfter}`} />}
          {isLoading && (
            <SpinnerContainer>
              <Spinner size="sm" />
            </SpinnerContainer>
          )}
        </InputContainer>
      </Label>

      {children}
      <TextHint isError={isError} isWarning={isWarning} isSuccess={isSuccess}>
        {hintMessage}
      </TextHint>
    </FieldGroup>
  );
};

export default InputForm;
export { TextHint };
