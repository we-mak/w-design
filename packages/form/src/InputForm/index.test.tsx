import * as React from "react";
import { cleanup, render } from "@testing-library/react";
import { Provider } from "@w-design/core";
import InputForm from "./index";

afterEach(cleanup);

describe("<InputForm/>", () => {
  it("should render correctly", () => {
    const { container } = render(
      <Provider>
        <InputForm label="label" hintMessage="hint" iconBefore="fas fa-phone" />
        <InputForm iconAfter="fas fa-phone" />
        <InputForm isError />
        <InputForm isFocus />
        <InputForm isDisabled />
        <InputForm isRequired />
        <InputForm isWarning />
        <InputForm isSuccess />
        <InputForm isLoading />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  // Props test
  it("should have label props", () => {
    const { getByText } = render(
      <Provider>
        <InputForm label="Hello" />
      </Provider>
    );
    getByText("Hello");
  });

  it("should have hint message", () => {
    const { getByText } = render(
      <Provider>
        <InputForm hintMessage="Hello" />
      </Provider>
    );
    getByText("Hello");
  });
});
