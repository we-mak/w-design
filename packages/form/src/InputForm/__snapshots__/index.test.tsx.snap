// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputForm/> should render correctly 1`] = `
<div>
  <div
    class="sc-jSFjdj hOedM"
  >
    <label
      class="sc-bdnxRM frNBIv"
    >
      label
       
      <span
        class="sc-dlnjwi drwbAa"
      >
        <i
          aria-hidden="true"
          class="Icon__StyledIcon-d8swi0-0 gNJswg icon-before fas fa-phone"
        />
        <input
          autocomplete="off"
          class="sc-gtsrHT bSXEP icon-before"
          type="text"
          value=""
        />
      </span>
    </label>
    <small
      class="sc-hKFxyN cBZnGA"
    >
      hint
    </small>
  </div>
  <div
    class="sc-jSFjdj hOedM"
  >
    <label
      class="sc-bdnxRM frNBIv"
    >
       
      <span
        class="sc-dlnjwi drwbAa"
      >
        <input
          autocomplete="off"
          class="sc-gtsrHT bSXEP icon-after"
          type="text"
          value=""
        />
        <i
          aria-hidden="true"
          class="Icon__StyledIcon-d8swi0-0 gNJswg icon-after fas fa-phone"
        />
      </span>
    </label>
    <small
      class="sc-hKFxyN cBZnGA"
    />
  </div>
  <div
    class="sc-jSFjdj hOedM"
  >
    <label
      class="sc-bdnxRM frNBIv"
    >
       
      <span
        class="sc-dlnjwi drwbAa"
      >
        <input
          autocomplete="off"
          class="sc-gtsrHT hkWwAO"
          type="text"
          value=""
        />
      </span>
    </label>
    <small
      class="sc-hKFxyN eiTtzt"
    />
  </div>
  <div
    class="sc-jSFjdj hOedM"
  >
    <label
      class="sc-bdnxRM frNBIv"
    >
       
      <span
        class="sc-dlnjwi drwbAa"
      >
        <input
          autocomplete="off"
          class="sc-gtsrHT bSXEP"
          type="text"
          value=""
        />
      </span>
    </label>
    <small
      class="sc-hKFxyN cBZnGA"
    />
  </div>
  <div
    class="sc-jSFjdj hOedM"
  >
    <label
      class="sc-bdnxRM frNBIv"
    >
       
      <span
        class="sc-dlnjwi drwbAa"
      >
        <input
          autocomplete="off"
          class="sc-gtsrHT bSXEP"
          disabled=""
          type="text"
          value=""
        />
      </span>
    </label>
    <small
      class="sc-hKFxyN cBZnGA"
    />
  </div>
  <div
    class="sc-jSFjdj hOedM"
  >
    <label
      class="sc-bdnxRM frNBIv"
    >
       
      <sup
        aria-hidden="true"
        class="sc-gKAaRy ggtRKD"
      >
        *
      </sup>
      <span
        class="sc-dlnjwi drwbAa"
      >
        <input
          autocomplete="off"
          class="sc-gtsrHT bSXEP"
          required=""
          type="text"
          value=""
        />
      </span>
    </label>
    <small
      class="sc-hKFxyN cBZnGA"
    />
  </div>
  <div
    class="sc-jSFjdj hOedM"
  >
    <label
      class="sc-bdnxRM frNBIv"
    >
       
      <span
        class="sc-dlnjwi drwbAa"
      >
        <input
          autocomplete="off"
          class="sc-gtsrHT jemeiw"
          type="text"
          value=""
        />
      </span>
    </label>
    <small
      class="sc-hKFxyN bIILMK"
    />
  </div>
  <div
    class="sc-jSFjdj hOedM"
  >
    <label
      class="sc-bdnxRM frNBIv"
    >
       
      <span
        class="sc-dlnjwi drwbAa"
      >
        <input
          autocomplete="off"
          class="sc-gtsrHT eOszqd"
          type="text"
          value=""
        />
      </span>
    </label>
    <small
      class="sc-hKFxyN hdWGPx"
    />
  </div>
  <div
    class="sc-jSFjdj hOedM"
  >
    <label
      class="sc-bdnxRM frNBIv"
    >
       
      <span
        class="sc-dlnjwi drwbAa"
      >
        <input
          autocomplete="off"
          class="sc-gtsrHT bSXEP"
          type="text"
          value=""
        />
        <span
          class="sc-eCApnc hjTAnw"
        >
          <div
            class="Styled__SpinnerWrapper-ues3u7-0 dzucZI"
          >
            <svg
              class="Styled__SpinnerCircular-ues3u7-1 eMdrEc"
              viewBox="25 25 50 50"
            >
              <circle
                class="Styled__SpinnerCircle-ues3u7-2 cxmtVQ"
                cx="50"
                cy="50"
                fill="none"
                r="20"
              />
            </svg>
          </div>
        </span>
      </span>
    </label>
    <small
      class="sc-hKFxyN cBZnGA"
    />
  </div>
</div>
`;
